File: 刷题/cpp/862. 和至少为 K 的最短子数组.cpp
class Solution {
public:
    int shortestSubarray(vector<int>& nums, int k) {
        deque<int> qu;
        int n=nums.size();
        long s[n+1];
        for(int i=0;i<n;i++){
            s[i+1]=s[i]+nums[i];
        }
        int ans=n+1;
        for(int i=0;i<=n;i++){
            long curnum=s[i];
            while(!qu.empty()&&curnum-s[qu.front()]>=k){
                ans=min(ans,i-qu.front());
                qu.pop_front();
            }
            while(!qu.empty()&&s[qu.back()]>=curnum){
                qu.pop_back();
            }
            qu.push_back(i);
        }
        return ans==n+1?-1:ans;
    }
};

File: 刷题/golang/862. 和至少为 K 的最短子数组.go
func shortestSubarray(nums []int, k int) int {
    n:=len(nums)
    preSum:=make([]int,n+1)
    for i, num:=range nums{
        preSum[i+1]=preSum[i]+num
    }
    ans:= n+1
    q:=[]int{}
    for i, curSum:=range preSum{
        for len(q)>0 && curSum-preSum[q[0]] >=k{
            ans=min(ans,i-q[0])
            q=q[1:]
        }
        for len(q)>0 && preSum[q[len(q)-1]]>=curSum{
            q=q[:len(q)-1]
        }
        q=append(q,i)
    }
    if ans<n+1{
        return ans
    }else {
        return  -1
    }
}
func min(a,b int) int{
    if a<b{
        return a
    }
    return b
}